// saber/dist/WebpackUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
class WebpackUtils {
    constructor(api) {
        this.api = api;
    }
    get shouldCache() {
        return this.api.config.build.cache !== false;
    }
    getCacheOptions(loader, obj) {
        return this.shouldCache
            ? {
                cacheDirectory: this.api.resolveCache(path_1.default.join('cache', loader)),
                cacheIdentifier: obj && JSON.stringify(typeof obj === 'function' ? obj() : obj)
            }
            : {};
    }
    addCacheSupport(rule, loader, obj) {
        if (this.shouldCache) {
            rule
                .use('cache-loader')
                .loader(require.resolve('cache-loader'))
                .options(this.getCacheOptions(loader, obj));
        }
    }
}
exports.WebpackUtils = WebpackUtils;
